providers/class.two-factor-fido-u2f.php

```diff
@@ -55,7 +147,190 @@ public function validate_authentication( $user ) {}
+				|| ! property_exists( $register, 'counter' ) || ( -1 > $register->counter )
+		) {
+			return false;
+		}
+
+		$register = array(
+			'keyHandle'   => $register->keyHandle,
+			'publicKey'   => $register->publicKey,
+			'certificate' => $register->certificate,
+			'counter'     => $register->counter,
+		);
+
+		$register['name']      = 'New Security Key';
+		$register['added']     = current_time( 'timestamp' );
+		$register['last_used'] = $register['added'];
+
+		return add_user_meta( $user_id, self::REGISTERED_KEY_USER_META_KEY, $register );
+	}
+
+	/**
+	 * Retrieve registered security keys for a user.
+	 *
+	 * @since 0.1-dev
+	 *
+	 * @param int $user_id User ID.
+	 * @return array|bool Array of keys on success, false on failure.
+	 */
+	public static function get_security_keys( $user_id ) {
+		if ( ! is_numeric( $user_id ) ) {
+			return false;
+		}
+
+		$keys = get_user_meta( $user_id, self::REGISTERED_KEY_USER_META_KEY );
+		if ( $keys ) {
+			foreach ( $keys as $index => $key ) {
+				$keys[ $index ] = (object) $key;
+			}
+		}
+
+		return $keys;
+	}
+
+	/**
+	 * Update registered security key.
+	 *
+	 * Use the $prev_value parameter to differentiate between meta fields with the
+	 * same key and user ID.
+	 *
+	 * If the meta field for the user does not exist, it will be added.
+	 *
+	 * @since 0.1-dev
+	 *
+	 * @param int    $user_id  User ID.
+	 * @param object $data The data of registered security key.
+	 * @return int|bool Meta ID if the key didn't exist, true on successful update, false on failure.
+	 */
+	public static function update_security_key( $user_id, $data ) {
+		if ( ! is_numeric( $user_id ) ) {
+			return false;
+		}
+
+		if (
+			! is_object( $data )
+				|| ! property_exists( $data, 'keyHandle' ) || empty( $data->keyHandle )
+				|| ! property_exists( $data, 'publicKey' ) || empty( $data->publicKey )
+				|| ! property_exists( $data, 'certificate' ) || empty( $data->certificate )
+				|| ! property_exists( $data, 'counter' ) || ( -1 > $data->counter )
+		) {
+			return false;
+		}
+
+		$keys = get_user_meta( $user_id, self::REGISTERED_KEY_USER_META_KEY );
+		if ( $keys ) {
+			foreach ( $keys as $index => $key ) {
+				if ( $key['keyHandle'] === $data->keyHandle ) {
+					return update_user_meta( $user_id, self::REGISTERED_KEY_USER_META_KEY, (array) $data, $key );
+				}
+			}
+		}
+
+		return self::add_security_key( $user_id, $data );
+	}
+
+	/**
+	 * Remove registered security key matching criteria from a user.
+	 *
+	 * @since 0.1-dev
+	 *
+	 * @param int    $user_id   User ID.
+	 * @param string $keyHandle Optional. Key handle.
+	 * @return bool True on success, false on failure.
+	 */
+	public function delete_security_key( $user_id, $keyHandle = null ) {
+		global $wpdb;
+
+		if ( ! is_numeric( $user_id ) ) {
+			return false;
+		}
+
+		$user_id = absint( $user_id );
+		if ( ! $user_id ) {
+			return false;
+		}
+
+		$table = $wpdb->usermeta;
```

Comment:
Why is this function doing direct database queries, instead of using `delete_user_meta()` ?

https://github.com/WordPress/WordPress/blob/ce557062f4123d8513378cf415b4e8b612c33ccc/wp-includes/user.php#L1304-L1321

`delete_user_meta()` can delete only where the meta value matches the third passed in argument, am I missing something here?